### PR TITLE
[mlir] `im2col` & `l2cache` on cp.async.bulk.tensor.shared.cluster.global`

### PR DESCRIPTION
**Pull request by grypp**
PR adds support of `im2col` and `l2cache` to `cp.async.bulk.tensor.shared.cluster.global`. The Op is now supports all the traits of the corresponding PTX instruction. 

The current structure of this operation looks somewhat like below. The PR also simplifies types so we don't need to write obvious types after `:` anymore. 
```
nvvm.cp.async.bulk.tensor.shared.cluster.global
		%dest, %tmaDescriptor, %barrier,
		box[%crd0,%crd1,%crd2,%crd3,%crd4]
		im2col[%off0,%off1,%off2] 			<-- PR introduces
		multicast_mask = %ctamask
		l2_cache_hint = %cacheHint			<-- PR introduces
		: !llvm.ptr<3>, !llvm.ptr
```

### CODE REVIEW

**Comment by grypp** [comment]
@durga4github it would be great if you could review. 
(I could not add you as a reviewer)

File: mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td [hunk 1]
```diff
@@ -1404,20 +1404,34 @@ def NVVM_CpAsyncBulkTensorGlobalToSharedClusterOp :
...
+    3-dimensional. 
+
     The `multicastMask` operand is optional. When it is present, the Op copies
     data from global memory to shared memory of multiple CTAs in the cluster.
     Operand `multicastMask` specifies the destination CTAs in the cluster such 
     that each bit position in the 16-bit `multicastMask` operand corresponds to
-     the `nvvm.read.ptx.sreg.ctaid` of the destination CTA.     
+    the `nvvm.read.ptx.sreg.ctaid` of the destination CTA.     
 
+    The `l2CacheHint` operand is optinal, and it is used to specify cache
```
Comment (durga4github) [inline mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:1432]: Sorry for the nit, 
'optional' typo

File: mlir/test/Conversion/NVVMToLLVM/nvvm-to-llvm.mlir [hunk 2]
```diff
@@ -85,93 +85,111 @@ func.func @async_cp_zfill(%dst: !llvm.ptr<3>, %src: !llvm.ptr<1>, %cpSize: i32)
   return
 }
 
+// CHECK-LABEL: @tma_load_4d_all
```
Comment (durga4github) [inline mlir/test/Conversion/NVVMToLLVM/nvvm-to-llvm.mlir:88]: can we add just one more of this _all test for 3d as well?

Then, that will cover all cases where im2col is valid + other optional operands.

File: mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td [hunk 3]
```diff
@@ -1404,20 +1404,34 @@ def NVVM_CpAsyncBulkTensorGlobalToSharedClusterOp :
   AttrSizedOperandSegments]>,
   Arguments<(ins  LLVM_PointerShared:$dstMem,
                   LLVM_AnyPointer:$tmaDescriptor,
-                  LLVM_PointerShared:$mbar,
-                  Optional<I16>:$multicastMask,                  
                   Variadic<I32>:$coordinates,
+                  LLVM_PointerShared:$mbar,                  
+                  Variadic<I16>:$im2colOffsets,
```
Comment (durga4github) [inline mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:1409]: I thought this 'im2colOffsets' would be an Optional operand too? No?

Comment (grypp) [reply mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:1409]: Yes you are right, `im2colOffset` is optional. We have `Variadic<I16>` that can be  0 sized. 

For example the following Op is valid and executes in `tiled` mode:
```
nvvm.cp.async.bulk.tensor.shared.cluster.global %0, %1, %2, 
                box[%d0,%d1,%d2] : !llvm.ptr<3>, !llvm.ptr
```
`im2col` mode is activated when offsets are present like below:
```
nvvm.cp.async.bulk.tensor.shared.cluster.global %0, %1, %2, 
                box[%d0,%d1,%d2] im2col[%off0]: !llvm.ptr<3>, !llvm.ptr
```

Comment (durga4github) [reply mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:1409]: Looking for your thoughts on this one, Otherwise, it looks good to me.

Comment (durga4github) [reply mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:1409]: ok, I was thinking we may need to provide an empty-vector like construct to take care of 0-sized im2col.
Not needing to provide that is even better !

+1. This change looks good to me.

**Review by durga4github — APPROVED** [review]
Looks good to me.

**Review by qcolombet — APPROVED** [review]

**Pull request merged**